pkg/ingester/owned_series_test.go

```diff
@@ -127,133 +151,471 @@ func TestOwnedSeriesService(t *testing.T) {
+				c.checkUpdateReasonForUser(t, "")
+
+				// add an ingester
+				c.registerSecondIngesterOwningHalfOfTheTokens(t)
+
+				// wait for the ingester to see the updated ring
+				test.Poll(t, 2*c.ringHeartbeatPeriod, 2, func() interface{} {
+					return c.ing.lifecycler.InstancesCount()
+				})
+
+				// pass ringChanged=true to trigger recompute
+				c.updateOwnedSeriesAndCheckResult(t, true, 1, recomputeOwnedSeriesReasonRingChanged)
+				c.checkUpdateReasonForUser(t, "")
+
+				// initial state: all series are owned by the second ingester
+				c.checkTestedIngesterOwnedSeriesState(t, 0, 1, ownedServiceTestUserSeriesLimit)
+
+				// change shard size to 2, splitting the series between ingesters
+				limits[ownedServiceTestUser].IngestionTenantShardSize = 2
+
+				c.updateOwnedSeriesAndCheckResult(t, false, 1, recomputeOwnedSeriesReasonShardSizeChanged)
+				c.checkTestedIngesterOwnedSeriesState(t, ownedServiceSeriesCount/2, 2, ownedServiceTestUserSeriesLimit/2)
+				c.checkUpdateReasonForUser(t, "")
+
+				// change shard size back to 1, moving the series back to the original ingester
+				limits[ownedServiceTestUser].IngestionTenantShardSize = 1
+
+				c.updateOwnedSeriesAndCheckResult(t, false, 1, recomputeOwnedSeriesReasonShardSizeChanged)
+				c.checkTestedIngesterOwnedSeriesState(t, 0, 1, ownedServiceTestUserSeriesLimit)
+				c.checkUpdateReasonForUser(t, "")
+			},
 		},
-
-		"early compaction removes all series": func(t *testing.T, c *ownedSeriesTestContext) {
-			c.pushUserSeries(t)
-
-			c.updateOwnedSeriesAndCheckResult(t, true, 1, recomputeOwnedSeriesReasonNewUser)
-			c.checkUserSeriesOwnedAndShardsByTestedIngester(t, ownedServiceSeriesCount, 0)
-
-			// Run early compaction. This removes all series from memory.
-			c.ing.compactBlocks(context.Background(), true, time.Now().Add(1*time.Minute).UnixMilli(), nil)
-			require.Equal(t, uint64(0), c.db.Head().NumSeries())
-
-			c.updateOwnedSeriesAndCheckResult(t, false, 1, recomputeOwnedSeriesReasonEarlyCompaction)
-			c.checkUserSeriesOwnedAndShardsByTestedIngester(t, 0, 0)
+		"unchanged ring and shards, series limit up and down": {
+			limits: map[string]*validation.Limits{
+				ownedServiceTestUser: {
+					MaxGlobalSeriesPerUser:   ownedServiceTestUserSeriesLimit,
+					IngestionTenantShardSize: 0,
+				},
+			},
+			testFunc: func(t *testing.T, c *ownedSeriesTestContext, limits map[string]*validation.Limits) {
+				c.pushUserSeries(t)
+				c.updateOwnedSeriesAndCheckResult(t, false, 1, recomputeOwnedSeriesReasonNewUser)
+				c.checkUpdateReasonForUser(t, "")
+
+				// initial state: all series are owned by the first ingester
+				c.checkTestedIngesterOwnedSeriesState(t, ownedServiceSeriesCount, 0, ownedServiceTestUserSeriesLimit)
+
+				// increase series limit
+				limits[ownedServiceTestUser].MaxGlobalSeriesPerUser = ownedServiceTestUserSeriesLimit * 2
+
+				c.updateOwnedSeriesAndCheckResult(t, false, 1, recomputeOwnedSeriesReasonLocalLimitChanged)
```

Comment:
I agree this isn't ideal, but it felt cleaner to have all of the owned series "state" updated in one place inside of `userDB.recomputeOwnedSeriesWithComputeFn`.

Practically, we tend to update the series limit and shard size at the same time (to keep a tenant's series/shard roughly constant), and the shard size change will trigger a recomputation anyway.

We can change this behavior if we really want to, but due to the complexity of juggling all the different recomputation reasons, I'd prefer to do it in a follow-up PR.  We also run recomputation on EVERY shard size change, even if the count is larger than the ingester count, so if we do end up changing the limit logic, we should probably fix that too.

Reply:
Agree to keep as-is in this PR. The only possible change I can imagine is to skip token ranges and owned series recomputation in some edge cases, but it would also make code more complex. Probably not worth it.